FR #962: Adding Correction Handling attribute

ijurica opened the pull request:
Note: This PR addresses part of the scope defined in FR #962.

---

This PR adds a new Correction Handling attribute to explicitly indicate how corrections should be represented in datasets. 
It covers various invoicing statuses of the original records and different data provisioning approaches, including replacement-style and append-only.

ijurica committed:
FOCUS #962: Correction Handling attribute spec

ijurica committed:
FOCUS #962: Correction Handling attribute spec

shawnalpay commented:
@ijurica Does this PR replace #1151, either totally or in part?

shawnalpay commented on supporting_content/attributes/correction_handling.md:
```diff
@@ -0,0 +1,72 @@
+
+## Cost and Usage data Provisioning Styles
+
+> **TODO:** Mention/address in upcoming Correction Handling attribute
```
Need to scrub `TODO`s.

shawnalpay commented on supporting_content/attributes/correction_handling.md:
```diff
@@ -0,0 +1,72 @@
+
+### Normative Requirements for Post-Invoice Finalization Corrections
+
+> **TODO:** Mention/address in upcoming Correction Handling attribute
```
Need to scrub `TODO`s.

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,88 @@
+
+## Requirements
+
+> WORK IN PROGRESS !!!
```
Need to scrub this line if this PR is truly out of draft.

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,88 @@
+
+## Exceptions
+
+> WORK IN PROGRESS !!!
```
Need to scrub this line if this PR is truly out of draft.

ijurica committed:
FOCUS #962: Correction Handling attribute spec

ijurica committed:
FOCUS #962: Correction Handling attribute spec

shawnalpay commented:
Discussed in Aug 19 TF1.  Put into draft.  @ijurica will complete and bring out of draft by Thu Aug 21.

ijurica committed:
FOCUS #962: Correction Handling attribute spec update - work in progress

ijurica committed:
FOCUS #962: Correction Handling attribute spec update - work in progress

jasonwucp commented:
Do you have a list of correction events that are currently handled through past finished month FOCUS cost and usage data restatements instead of using correction line items in the new month? One example I can think of is the out-of-cycle billing for one-time purchases that come after month-end, where the finished month data needs to be restated to reflect it.

shawnalpay committed:
linter fix

shawnalpay committed:
linter fix

ijurica committed:
FOCUS #962: Correction Handling attribute spec update - work in progress

ijurica committed:
FOCUS #962: Correction Handling attribute spec update - work in progress

ijurica committed:
FOCUS #962: Correction Handling attribute spec update - work in progress

ijurica committed:
FOCUS #962: Correction Handling attribute spec update  - supporting content scrub

ijurica committed:
FOCUS #962: Correction Handling attribute spec update  - supporting content scrub

ijurica committed:
FOCUS #962: Correction Handling attribute - ignore

ijurica committed:
FOCUS #962: Correction Handling attribute - ignore

ijurica committed:
FOCUS #962: Correction Handling attribute spec update

ijurica committed:
FOCUS #962: Correction Handling attribute spec update

ijurica committed:
FOCUS #962: Correction Handling attribute supporting content scrub

ijurica committed:
FOCUS #962: Correction Handling attribute supporting content scrub

shawnalpay reviewed (changes requested):
@ijurica 
I would expect to see use of this attribute in the spec -- otherwise, it's just sitting there doing nothing 😅 

Could you please leverage the attribute via normative requirement for at least one column?

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,95 @@
+* Billing period MUST be considered invoiced and closed once all invoices for that period are issued.
+* Additional charge records MUST NOT be associated with an invoice once it is issued.
+* Additional charge records MUST NOT be associated with a billing period once it is invoiced and closed.
+
```
These spaces make the entire bullet-point list get vertically spaced.  You may want to add subheaders if you want to delineate between sections -- otherwise the formatting gets weird.

ijurica replied:
added subheaders. resolving.

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,95 @@
+# Correction Handling
```
I think that this write-up is long enough that you need to introduce some subheaders for ease of reading.

ijurica replied:
addressed. resolving.

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,95 @@
+
+A billing period is considered invoiced (or closed) once all invoices for that period have been issued and all charge records for that period are finalized. After a billing period is invoiced, no new charge records may be associated with it, and all previously finalized charge records remain unchanged. Any necessary corrections to charges originally incurred in an invoiced billing period must ins ...
+
+FOCUS supports two cost and usage data delivery mechanisms: Replacement and Append-only.
```
I suggest we bold these terms.
```suggestion
FOCUS supports two cost and usage data delivery mechanisms: **Replacement** and **Append-only**.
```

ijurica replied:
I’ve added subheaders. It seems to me that bolding is no longer critical (since we haven’t used bolding anywhere else). 
@shawnalpay If you feel we should still keep bolding, feel free to reopen the conversation.

rileyjenk commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,95 @@
+
+A billing period is considered invoiced (or closed) once all invoices for that period have been issued and all charge records for that period are finalized. After a billing period is invoiced, no new charge records may be associated with it, and all previously finalized charge records remain unchanged. Any necessary corrections to charges originally incurred in an invoiced billing period must ins ...
+
+FOCUS supports two cost and usage data delivery mechanisms: Replacement and Append-only.
```
There is an assumption here that isn't accurate. For many billing systems there exists the presence of a "we update the data to meet our customer needs" and that can take many forms. For example a vendor may do a partition update where their typical update pattern is to append additional partitions and update existing partitions.

ijurica replied:
I believe the described scenario falls into Replacement mechanism, which doesn't exclude appending.

shawnalpay replied:
Discussed in Sep 2 TF2.  @ijurica to mention this scenario under `Replacement`.

ijurica replied:
The introductory section has been updated. I hope this now covers the hybrid scenario described above. 

> FOCUS supports two cost and usage data delivery mechanisms: Replacement and Append-only. **These mechanisms are not mutually exclusive and hybrid implementations are common, allowing Data Generators to meet specific delivery and auditability requirements.**

---

Resolving the conversation for now. Feel free to reopen and provide suggestions if additional adjustments are needed.

---

Note: Softening the language related to what’s permitted in invoiced billing periods is a separate conversation.

rileyjenk commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,95 @@
+* Cost allocation and chargeback - attributing corrections to the correct dimensions (e.g., Billing Account, Sub Account, SKU ID, SKU Price ID, Resource ID).
+* Auditability - tracing the full lifecycle of a charge from the original record through all subsequent corrections.
+
+Once an invoice is issued, it serves as the authoritative financial document and is considered finalized and immutable. All charge records associated with an issued invoice are also considered finalized and must remain unchanged (i.e., corrections to finalized charge records, whether as updates, deletions or omissions, are not permitted). Furthermore, no additional charge records may be associate ...
```
This is speaking in terms of the invoice being unchanged, which is different that the Cost and Usage Dataset being unchanged.

ijurica replied:
I’d like to discuss the so-called Corrections to Cost vs Corrections to Usage topic that @rileyjenk mentioned last week outside the TF-1 meeting. As I understand it, this is about data corrections that don’t impact invoices. 

It would be great if @rileyjenk  could elaborate a bit more on that...

shawnalpay replied:
Discussed in Sep 2 TF1.  @rileyjenk will provide @ijurica some examples that do not fit the framing as laid out here; Irena will then revise the verbiage to handle.  Irena requests that these examples include how to denote Billing Period Start and End.

ijurica replied:
See scenarios S-1 and S-1-B in this [sheet](https://docs.google.com/spreadsheets/d/1UDZCxPqUNEUQt90h8sW-YuhgBsk4pHYcwRlgPJVmwPo/edit?gid=1769206673#gid=1769206673).

ijurica replied:
These paragraphs have been reworded - attaching the current version below:

> Once an invoice is issued, it becomes the authoritative financial document. Issued invoice is considered finalized, and the financial data it contains must not be altered. While modifications to the underlying cost and usage charges associated with an issued invoice (e.g., updates, additions, or omissions) may be permitted, they must not compromise the integrity of the issued invoice. Only modifications that maintain alignment with the invoice content are acceptable. Any misalignment would invalidate the prior *invoice reconciliation* and undermine the invoice's financial validity.
> 
> Corrections to the underlying cost and usage charges associated with an issued invoice that do not impact data presented on the invoice are allowed. However, although these modifications do not affect *invoice reconciliation*, they can still result in loss of auditability and traceability, which in turn complicates modifications and mappings required in downstream FinOps activities, such as cost allocation, chargeback, or budgeting. For this reason, such modifications are not preferred and should only be applied when explicitly requested by the end-user.

Also, [here](https://docs.google.com/spreadsheets/d/1UDZCxPqUNEUQt90h8sW-YuhgBsk4pHYcwRlgPJVmwPo/edit?gid=1769206673#gid=1769206673) you can find both the scenario and the sample data that were used as the basis for the discussion.

I’m closing this conversation, but if you feel the concern still hasn’t been properly addressed, please post suggestion or feedback on the latest version or feel free to reopen this thread.

rileyjenk commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,95 @@
+* Cost allocation and chargeback - attributing corrections to the correct dimensions (e.g., Billing Account, Sub Account, SKU ID, SKU Price ID, Resource ID).
+* Auditability - tracing the full lifecycle of a charge from the original record through all subsequent corrections.
+
+Once an invoice is issued, it serves as the authoritative financial document and is considered finalized and immutable. All charge records associated with an issued invoice are also considered finalized and must remain unchanged (i.e., corrections to finalized charge records, whether as updates, deletions or omissions, are not permitted). Furthermore, no additional charge records may be associate ...
```
This does not work. What if the invoice issuer has $100 on it but the Cost and usage data only has $90 because an entry was mistakenly not populated in the data. In order to fix it I need the ability to add a row that was excluded. There was no financial correction in a followup period. I got invoiced, it just wasn't in the Cost and Usage dataset.

ijurica replied:
I'd argue this falls under: Exceptions … due to technical issues encountered during or after invoice issuance or billing period closure.

I agree that that exception is rather vague. I could try to rephrase it to make it more specific, but I’m hesitant in case I leave something out. 😞

rileyjenk commented:
I believe our thinking around correction handling needs to fundamentally change. Our correction handling need to facilitate DataGenerators to correct their FOCUS datasets regardless of update mechanism to how they populate their data.  Specifying when they can and can't change their FOCUS datasets will make it more difficult for them to handle unfun scenarios when the stuff has hit the fan and data is wrong. 

As it reads today I would chalk everything up to technical issue and make any change I need to get the data correct

I would structure it into 3 different correction areas/rulesets:

- Corrections for Costs: Error impacts Billed Cost 
- - Billed Cost: 
- - - Closed Periods: Corrections where charges where not included or refund is required happen in subsequent invoice periods. These may require changes to records in Cost and Usage for already issued invoices to fix billed cost for existing row or inject missing row. It may also require new rows in current cost and usage invoice period to show the invoice charges in the subsequent invoice to show refund or extra charges. 
- - - Open Periods: Allow data generators to make any correction needed to ensure these numbers are accurate
- Corrections for Usage and EffectiveCost: : Allow data generators to make any correction needed to ensure these numbers are accurate
- Corrections for entries that are not cost or usage: 
- - The error has no impact to usage amounts or cost amounts (for example wrong value for a service name), allow any correction.

ijurica committed:
FOCUS #962: Correction Handling attribute - adding subtitles

ijurica committed:
FOCUS #962: Correction Handling attribute - adding subtitles

ijurica committed:
FOCUS #962: Correction Handling attribute - adding subtitles

ijurica committed:
FOCUS #962: Correction Handling attribute - adding subtitles

ijurica committed:
FOCUS #962: Correction Handling attribute - adding subtitles

ijurica committed:
FOCUS #962: Correction Handling attribute - adding subtitles

ijurica committed:
FOCUS #962: Correction Handling attribute - requirements breakdown

ijurica committed:
FOCUS #962: Correction Handling attribute - requirements breakdown

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+All corrections adhere to the following requirements:
+
+### Invoice and Billing Period Requirements
```
Discussed in Sep 2 TF1.  @ijurica will move this section into a separate `Invoice Handling` markdown -- either here in this PR, or in the parallel-path PR #1082.

ijurica replied:
Invoice- and billing period-related normative requirements have been relocated from Correction Handling to Invoice Handling, and the headers in the Correction Handling Requirements section have been removed.

---
resolving this conversation

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+* Additional *charges* MUST NOT be associated with an invoice once it is issued.
+* Additional *charges* MUST NOT be associated with a billing period once it is invoiced and closed.
+
+### General Requirements for Corrections
```
Discussed in Sep 2 TF1.  @ijurica will strike the headers in the Requirements section.

ijurica replied:
Invoice- and billing period-related normative requirements have been relocated from Correction Handling to Invoice Handling, and the headers in the Correction Handling Requirements section have been removed.

---
resolving this conversation

shawnalpay reviewed (changes requested):
Please address latest comments from me and @rileyjenk, per discussion in Sep 2 TF1.

ijurica committed:
FOCUS #962: correction-handling-attribute - Merge remote-tracking branch 'origin/working_draft' into 962-correction-handling-attribute

flanakin commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+### Invoice and Billing Period Requirements
+
+* Invoice MUST be considered finalized and immutable once issued.
```
How can we have an invoice requirement in a cost dataset? That needs to be a requirement on the invoice itself. We can't enforce that here. Unless this moves to an "Assumptions" section.

ijurica replied:
resolving in favor of [this conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368/files#r2352274007)

flanakin commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+* Invoice MUST be considered finalized and immutable once issued.
+* Once the associated invoice is issued, each underlying *charges* adheres to the following additional requirements:
+  * *Charge* MUST be considered finalized and immutable.
```
This also feels like an assumption. (Although the next one is definitely a requirement on the dataset.)

ijurica replied:
resolving in favor of [this conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368/files#r2352274007)

flanakin commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+  * *Charge* MUST NOT be updated, deleted, or omitted.
+* Billing period MUST be considered invoiced and closed once all invoices for that period are issued.
+* Additional *charges* MUST NOT be associated with an invoice once it is issued.
+* Additional *charges* MUST NOT be associated with a billing period once it is invoiced and closed.
```
nit: Should these be requirements on the InvoiceId and BillingPeriod* columns? I'm on the fence whether these are assumptions or hard requirements we can hold data generators accountable for.

shawnalpay replied:
There are certainly some requirements here that belong in an Invoice Handling attribute, which is forthcoming per @Matt-Cowsert and PR #1321.

Hopefully we can get that PR merged this week, and then @ijurica can propose changes.  Otherwise, she'll need to stand up a separate Invoice Handling here in this PR, and then we'll have to reconcile later.

ijurica replied:
resolving in favor of [this conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368/files#r2352274007)

flanakin commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+### Constraints in an Invoiced Billing Period
+
+* Corrections to charges from a previously invoiced and closed billing period adhere to the following additional requirements:
+  * ChargeClass MUST be "Correction".
+  * Correction MUST NOT replace or omit the original record.
+  * Corrected row(s) MUST be assigned to a different InvoiceId than the original record.
+  * [BillingPeriodStart](#billingperiodstart) and [BillingPeriodEnd](#billingperiodend) MUST equal the [*inclusive start bound*](#glossary:inclusivestartbound) and [*exclusive end bound*](#glossary:exclusiveendbound) of a subsequent open billing period in which the correction is issued.
+  * [ChargePeriodStart](#chargeperiodstart) and [ChargePeriodEnd](#chargeperiodend) MUST equal the *inclusive start bound* and *exclusive end bound* of the period in which the cost was originally incurred.
```
```suggestion
Corrections to charges from a previously invoiced and closed billing period adhere to the following additional requirements:

* ChargeClass MUST be "Correction".
* Correction MUST NOT replace or omit the original record.
* Corrected row(s) MUST be assigned to a different InvoiceId than the original record.
* [BillingPeriodStart](#billingperiodstart) and [BillingPeriodEnd](#billingperiodend) MUST equal the [*inclusive start bound*](#glossary:inclusivestartbound) and [*exclusive end bound*](#glossary:exclusiveendbound) of a subsequent open billing period in which the correction is issued.
* [ChargePeriodStart](#chargeperiodstart) and [ChargePeriodEnd](#chargeperiodend) MUST equal the *inclusive start bound* and *exclusive end bound* of the period in which the cost was originally incurred.
```

ijurica replied:
Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

flanakin commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+### Constraints in the Replacement Mechanism
+
+* Replacement mechanism adheres to the following additional requirements:
+  * Correction handling implementation SHOULD support optional external retention of historical snapshots to enable traceability in uninvoiced billing periods (including both previous and current).
+  * Corrections for previously invoiced billing periods MUST be represented exclusively through the addition of new records.
+  * Corrections for uninvoiced billing periods MAY include updates, additions, or omissions.
```
```suggestion
Replacement mechanism adheres to the following additional requirements:

* Correction handling implementation SHOULD support optional external retention of historical snapshots to enable traceability in uninvoiced billing periods (including both previous and current).
* Corrections for previously invoiced billing periods MUST be represented exclusively through the addition of new records.
* Corrections for uninvoiced billing periods MAY include updates, additions, or omissions.
```

ijurica replied:
Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

flanakin commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+### Constraints in the Append-only Mechanism
+
+* Append-only mechanism adheres to the following additional requirements:
+  * All previously delivered records MUST be retained without modification or deletion.
+  * All corrections MUST be represented exclusively by adding new records.
```
```suggestion
Append-only mechanism adheres to the following additional requirements:

* All previously delivered records MUST be retained without modification or deletion.
* All corrections MUST be represented exclusively by adding new records.
```

ijurica replied:
Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

flanakin commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+## Exceptions
+
+* Exceptions to the restrictions on issued invoices, invoiced billing periods, and finalized charge records MAY apply in the following cases:
+  * Upon explicit request from the end-user (subject to validation and approval processes).
+  * Due to technical issues encountered during or after invoice issuance or billing period closure.
```
```suggestion
Exceptions to the restrictions on issued invoices, invoiced billing periods, and finalized charge records MAY apply in the following cases:

* Upon explicit request from the end-user (subject to validation and approval processes).
* Due to technical issues encountered during or after invoice issuance or billing period closure.
```

ijurica replied:
Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

SatyaVandrangi1 commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+Accurate correction handling is essential for a range of business-critical processes, including but not limited to:
+
+* Temporal accuracy - capturing both when the charge was incurred (reflected in charge period columns, i.e., Charge Period Start and Charge Period End) and when the correction was invoiced (reflected in billing period columns, i.e., Billing Period Start and Billing Period End columns).
+* Financial and legal integrity - preserving original charge records associated with finalized invoices while recording changes to those records separately, as finalized invoices represent binding financial documents requiring immutability and traceability.
```
I am not sure how strong a word immutability is. It is a valid scenario to completely cancel an issued invoice and re-issue a new invoice without replacing cost and usage records. As long as that scenario is considered immutable - we are good

shawnalpay replied:
Mentioning / incorporating exceptions around immutability is important.

shawnalpay replied:
Discussed in Sep 16 TF1.  Will close in favor of the file-level comment talking about how to resolve immutability.

SatyaVandrangi1 commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+
+Once an invoice is issued, it serves as the authoritative financial document and is considered finalized and immutable. All charge records associated with an issued invoice are also considered finalized and must remain unchanged (i.e., corrections to finalized charge records, whether as updates, deletions or omissions, are not permitted). Furthermore, no additional charge records may be associate ...
+
+A billing period is considered invoiced (or closed) once all invoices for that period have been issued and all charge records for that period are finalized. After a billing period is invoiced, no new charge records may be associated with it, and all previously finalized charge records remain unchanged. Any necessary corrections to charges originally incurred in an invoiced billing period must ins ...
```
There might be more to this. For a scenario, where the billing period is invoiced but the customer has not paid for it yet. Sometimes that customers might prefer to restate the invoice for the old billing period itself. Not replace but restate with a new invoice id. I believe this statement would not allow for that flexibility - since it says "must instead be reflected in a subsequent open billing period"

ijurica replied:
The exception had already been added to the Exception section, and in the current version it’s also included in this intro section 

> Exceptionally, additional invoices may be issued for a closed billing period only if explicitly requested by the end-user.

I hope this covers the scenario described here, so I’m closing this conversation. That said, if you feel the concern hasn’t been fully addressed, please post new feedback on the latest version or feel free to reopen this thread.

SatyaVandrangi1 commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,121 @@
+  * *Charge* MUST NOT be updated, deleted, or omitted.
+* Billing period MUST be considered invoiced and closed once all invoices for that period are issued.
+* Additional *charges* MUST NOT be associated with an invoice once it is issued.
+* Additional *charges* MUST NOT be associated with a billing period once it is invoiced and closed.
```
same as my comment above. Sometimes customers prefer the changes done to existing billing period if the payment was done yet

ijurica replied:
As I mentioned earlier in [this ](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368#discussion_r2399569914)thread, in the Invoice Handling - Exception section we specify:

> * Exceptions to the restrictions on *issued invoices* and *closed billing period* MAY apply in the following cases:
>   * Upon explicit request from the end-user (subject to validation and approval processes).

Regarding the individual normative requirements, although for a while were mentioning the exception directly in relevant requirements, after several discussions with TF-1 members and maintainers, we landed on keeping it in the Exception section without explicitly mentioning it in a particular requirement - at least for now (the same approach applied in other attribute definitions).

With that, I’m closing this conversation. Of course, please feel free to reopen it if you feel this needs further discussion.

shawnalpay commented on specification/attributes/correction_handling.md:
Discussion in Sep 9 TF1: we need to be more clear what "immutable" means.

ijurica replied:
Immutability is mentioned in several places within the Correction Handling and Invoice Handling attributes. It is currently tied to issued invoices and finalized charge records.

I think we should maintain immutability for issued invoices, but for finalized charge records we have a couple of options:

* Distinguish invoice-relevant vs non-invoice-relevant dimensions/metrics:
  * Apply stronger immutability restrictions to elements that appear on the invoice.
  * The challenge is identifying exactly which dimensions and metrics are present on the invoice.
* Tie immutability only to the billing period and billed cost:
  * In this approach, the content of the invoice and the dataset artifact may differ.

Consequently, stronger restrictions would apply to immutable elements within charge records, while for other elements, we could suggest that a Replacement approach is allowed if the user chooses.

Still not sure how to account for technical issues (see the example @rileyjenk mentioned in [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368#discussion_r2310555964) )

shawnalpay replied:
Discussed in Sep 16 TF1.  @ijurica will start with a list of proposed immutable columns.  We would at a minimum start with:

- Invoice ID
- Billing Period Start
- Billing Period End
- Billed Cost

Immutability here would be that they must match what's listed on the invoice -- and there may be some technical problem with the FOCUS dataset that may need to be fixed, which would require it being mutable _in order to get the dataset and the invoice to match_.

ijurica replied:
We no longer use the term immutable, even in the context of issued invoice descriptions. For example, rather than saying that "*...an issued invoice is considered immutable...*", the current version includes the following Invoice Handling normative requirements, relevant for this conversation:

> * *Invoice reconciliation* MUST include (but is not limited to) the following metrics and dimensions: BilledCost, BillingCurrency, InvoiceId, InvoiceIssuer, BillingAccountId, BillingPeriodStart, and BillingPeriodEnd.
> * *Invoice reconciliation* MUST be performed by the Invoice Issuer before issuing an invoice.
> * Cost and usage data presented on an invoice and included in *invoice reconciliation* MUST be documented by the Invoice Issuer and accessible to practitioners.
> * Cost and usage data presented on an *issued invoice* MUST NOT be altered.

Regarding charge records, modifications (updates, additions, or omissions) are allowed, but under specific conditions:

> * Correction to *issued charges* adheres to the following additional requirements:
>   * Correction MUST NOT be applied when it results in discrepancies with the cost and usage data presented on the associated *issued invoice*.
>   * Correction SHOULD NOT be applied when it does not results in discrepancies with the cost and usage data presented on the associated *issued invoice*, but still affects downstream FinOps capabilities (e.g., chargeback).
>   * Correction MAY include additional charges associated with a different InvoiceId than the original charges.
>   * Correction MAY include additional charges associated with a different billing period than the original charges.
> * Correction to previously *closed billing period* that requires issuing additional invoices MUST result in additional charge(s) associated with a subsequent *open billing period*, with the charge period indicating when the cost was incurred.
> 

---

*Note: I would primarily like to get feedback on the proposed concept. Of course, I’m open to suggestions on how to rephrase.*

shawnalpay commented:
Putting this PR back in draft per @ijurica's request.  She will be prepared to bring it out of draft with the necessary changes by next Tue Sep 16 TF1.

ijurica committed:
Merge branch '962-correction-handling-attribute' of https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec into 962-correction-handling-attribute

# Conflicts:
#	specification/attributes/correction_handling.md

ijurica committed:
FOCUS #962: correction-handling-attribute - Merge remote-tracking branch 'origin/working_draft' into 962-correction-handling-attribute

ijurica committed:
FOCUS #962: correction-handling-attribute - temp minimum alignment with matt_cowsert/1082-invoice-handling-requirements

ijurica committed:
FOCUS #962: correction-handling-attribute - cost and usage dataset requirements update

ijurica committed:
FOCUS #962-correction-handling-attribute: relocating AS-IS invoice and billing period content from correction-handling to invoice-handling

ijurica commented:
follow-up on this comment:

> @ijurica I would expect to see use of this attribute in the spec -- otherwise, it's just sitting there doing nothing 😅
> 
> Could you please leverage the attribute via normative requirement for at least one column?

@shawnalpay we are now referencing the Correction Handling attribute in a Cost and Usage Dataset Requirement

> * CostAndUsage MUST conform to [CorrectionHandling](#correctionhandling) requirements.

ijurica committed:
FOCUS #962-correction-handling-attribute: Correction Handling attribute update

ijurica commented on specification/attributes/invoice_handling.md:
```diff
@@ -0,0 +1,41 @@
+* Billing period MUST be considered invoiced and closed once all invoices for that period are issued.
+* Once the associated invoice is issued, each underlying *charge* adheres to the following additional requirements:
+  * *Charge* MUST be considered finalized and immutable.
+  * *Charge* MUST NOT be updated, deleted, or omitted.
```
Consolidating several of @flanakin’s comments that were originally posted on these requirements before they were relocated to Invoice Handling.

> How can we have an invoice requirement in a cost dataset? That needs to be a requirement on the invoice itself. We can't enforce that here. Unless this moves to an "Assumptions" section.
> 
> * Invoice MUST be considered finalized and immutable once issued.
>

I don't think this is an Assumption.

> This also feels like an assumption. (Although the next one is definitely a requirement on the dataset.)
> 
> * Invoice MUST be considered finalized and immutable once issued.
> * Once the associated invoice is issued, each underlying *charges* adheres to the following additional requirements:
>   * *Charge* MUST be considered finalized and immutable.
>   

We might soften the language here or consider distinguishing between immutable and mutable dimensions/metrics.

> nit: Should these be requirements on the InvoiceId and BillingPeriod* columns? I'm on the fence whether these are assumptions or hard requirements we can hold data generators accountable for.
>
> * Additional *charges* MUST NOT be associated with an invoice once it is issued.
> * Additional *charges* MUST NOT be associated with a billing period once it is invoiced and closed.
> 

I hope that as part of FR #1389 we will establish criteria to determine which requirements are dataset-level and which are column-level. For now, I suggest leaving these two requirements on the Invoice Handling attribute.

ijurica replied:
Follow-up feedback:

### 1. Comment

> How can we have an invoice requirement in a cost dataset? That needs to be a requirement on the invoice itself. We can't enforce that here. Unless this moves to an "Assumptions" section.
> 
> * Invoice MUST be considered finalized and immutable once issued.
>

#### Resolution: Moved the requirement to Invoice Handling and rephrased as:

* Cost and usage data presented on an *issued invoice* MUST NOT be altered.

### 2. Comment

> This also feels like an assumption. (Although the next one is definitely a requirement on the dataset.)
> 
> * Invoice MUST be considered finalized and immutable once issued.
> * Once the associated invoice is issued, each underlying *charges* adheres to the following additional requirements:
>   * *Charge* MUST be considered finalized and immutable.
>

#### Resolution: Rephrased as:

* Correction to *issued charges* adheres to the following additional requirements:
  * Correction MUST NOT be applied when it results in discrepancies with the cost and usage data presented on the associated *issued invoice*.
  * Correction SHOULD NOT be applied when it does not results in discrepancies with the cost and usage data presented on the associated *issued invoice*, but still affects downstream FinOps capabilities (e.g., chargeback).
  * Correction MAY include additional charges associated with a different InvoiceId than the original charges.
  * Correction MAY include additional charges associated with a different billing period than the original charges.

### 3. Comment

> nit: Should these be requirements on the InvoiceId and BillingPeriod* columns? I'm on the fence whether these are assumptions or hard requirements we can hold data generators accountable for.
>
> * Additional *charges* MUST NOT be associated with an invoice once it is issued.
> * Additional *charges* MUST NOT be associated with a billing period once it is invoiced and closed.
> 

#### Resolution: Replaced with  couple of requirements under Invoice Handling and Correction Handling:

Invoiced Handling:

* Billing period MUST be considered "open" until all planned invoices for that period have been issued.
* Billing period MUST be considered "closed" when all planned invoices for that period have been issued.
* Ability to determine whether a billing period is "open" or "closed" MUST be documented by the Invoice Issuer and accessible to practitioners.
* Additional invoice MUST NOT be associated with a *closed billing period*.

Correction Handling:

* Correction to previously *closed billing period* that requires issuing additional invoices MUST result in additional charge(s) associated with a subsequent *open billing period*, with the charge period indicating when the cost was incurred.

ijurica commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,96 @@
+* Corrections to charges from a previously invoiced and closed billing period adhere to the following additional requirements:
+  * ChargeClass MUST be "Correction".
+  * Correction MUST NOT replace or omit the original record.
+  * Corrected row(s) MUST be assigned to a different InvoiceId than the original record.
```
```suggestion
  * Corrected row(s) MUST be assigned to a different InvoiceId than the original record, unless a replacement approach is explicitly requested by the end-user.
```

ijurica committed:
FOCUS #962-correction-handling-attribute: relocating AS-IS Exceptions from correction-handling to invoice-handling

ijurica committed:
FOCUS #962-correction-handling-attribute: updates to invoice handling and glossary

ijurica committed:
FOCUS #962-correction-handling-attribute: updates to correction  handling

shawnalpay commented:
@ijurica Now that PR #1321 has been approved and merged into `working_draft`, there are a few conflicts between that work and yours, especially in `invoice_handling.md`.  Please review and resolve, and let me know if you want/need to huddle to resolve.  Thanks!

ijurica committed:
FOCUS #962-correction-handling-attribute: invoice_handling - temp. revert

ijurica committed:
FOCUS #962-correction-handling-attribute: Merge remote-tracking branch 'origin/working_draft' into 962-correction-handling-attribute

# Conflicts:
#	specification/attributes/attributes.mdpp
#	specification/datasets/cost_and_usage/dataset.md

ijurica committed:
FOCUS #962-correction-handling-attribute: invoice_handling update

ijurica committed:
FOCUS #962-correction-handling-attribute: Additional glossary terms.

ijurica committed:
FOCUS #962-correction-handling-attribute: glossary update

ijurica committed:
FOCUS #962-correction-handling-attribute: invoice-handling and correction-handling updates

shawnalpay commented on specification/glossary.md:
Change definition of Correction to include verbiage of "invoiced charge and/or closed billing period".

We may also need an entry for "invoiced charge".

ijurica replied:
### Correction 

- Rephrased and repurposed the glossary definition of Correction to reflect "correction" as a general concept,
- Added note to address confusion between the general concept (correction) and the specific Charge Class value - Correction (capitalized).

> <a name="glossary:correction"><b>Correction</b></a>
> 
> Any modification (including updates, additions, or omissions) to previously provided [*charges*](#glossary:charge), whether from an [*open billing period*](#glossary:open-billing-period) or a [*closed billing period*](#glossary:closed-billing-period). A correction may consist of multiple simultaneous modifications, each representing an update or omission of a previously delivered charge, or the addition of a new charge, collectively forming the correction.
> The term "Correction" (capitalized) refers specifically to an allowed value in the [Charge Class](#chargeclass) column, designating charges used to correct cost or usage data from a previously closed billing period. In contrast, the broader concept of "corrections" (lowercase) includes all types of modifications, regardless of the [*billing period*](#glossary:billing-period) status.

- **Note:** The allowed value "Correction" in the Charge Class column was not modified (at least not yet 😄), despite known naming ambiguity, due to prior consensus and sensitivity around this decision.

### Invoiced Charge
 
- Opted for the term Issued Charge instead of Invoiced Charge.

> <a name="glossary:issued-charge"><b>Issued Charge</b></a>
> 
> A [*charge*](#glossary:charge) associated with an [*issued invoice*](glossary:issued-invoice).

shawnalpay commented on specification/attributes/correction_handling.md:
```diff
@@ -0,0 +1,96 @@
+
+### Definition and Scope of Corrections
+
+Correction Handling attribute defines how updates to previously provided charge records are represented in FOCUS datasets.
```
Per our discussion in Sep 30 TF1: this is not necessarily accurate, per the point you made, @ijurica, around late-arriving charges.  I think this needs to become something like:

```suggestion
Correction Handling attribute defines how updates to invoiced charges and/or closed billing periods are represented in FOCUS datasets.
```

ijurica replied:
Although the previous discussion in TF-1 was focused on corrections to charges in previously *closed billing periods* and corrections to *invoiced charges* (charges associated with issued invoices), the Correction Handling attribute should address **corrections to all previously provided charges**, including those in the current billing period. 

For instance, we are describing that the append-only mechanism does not allow updating any previously provided charge, whereas the replacement mechanism does allow such updates. Therefore, I’d suggest we keep the current phrasing.

shawnalpay reviewed (changes requested):
@ijurica 
I started to go through the PR, but I think you first need to take a pass to make use of the terms:

 * *Invoiced charge*
 * *Closed billing period*
 
 ...as agreed in Sep 30 TF1.  If you make those changes, then I think this PR will be in much better shape, and I'll not drown you in comments/suggestions 😅 

Ping me once you've done so, and then I'll give it another look.

shawnalpay committed:
linter fix

ijurica committed:
FOCUS #962-correction-handling-attribute: renaming invoiced to closed billing period

ijurica committed:
FOCUS #962-correction-handling-attribute: adding open billing period glossary term

ijurica committed:
FOCUS #962-correction-handling-attribute: invoice handling and correction handling updates in progress

ijurica committed:
FOCUS #962-correction-handling-attribute: invoice handling and correction handling updates in progress

ijurica committed:
FOCUS #962-correction-handling-attribute: invoice handling updates

ijurica committed:
FOCUS #962-correction-handling-attribute updates

ijurica committed:
FOCUS #962-correction-handling-attribute: invoice-handling updates

ijurica committed:
FOCUS #962-correction-handling-attribute: clarification updates based on feedback

ijurica committed:
FOCUS #962-correction-handling-attribute: adding issued charge glossary term

ijurica committed:
FOCUS #962-correction-handling-attribute: Changing the Correction glossary term definition and (non-material) updates to ChargeClass column definition upon feedback

ijurica committed:
FOCUS #962-correction-handling-attribute: Correction glossary term update

ijurica committed:
FOCUS #962-correction-handling-attribute: moving corrections-related content from invoice handling to correction handling

ijurica committed:
linter fix

shawnalpay commented:
I duplicated the branch for this PR and issued PR #1501, which is a heavy refactor/revision of this PR.  I have partnered with @ijurica to address any concerns, and we will move forward with #1501 as a replacement for #1368.

Closing this PR, but not deleting the branch yet, as some of that work may be applicable to future work pertaining to #1500.
